src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceResponseTracker.cs

```diff
@@ -50,29 +51,31 @@ public override string ToString()
                 return "faulted";
             else
             {
-                return string.Format("[{0} {1}]",
-                     Task.Result.ToString(),
-                     (RemoteOwner.Address != null) ? RemoteOwner.Address.ToString() : "");
+                return
+                    $"[{Task.Result} {RemoteOwner.Address}]";
             }
         }
 
-
         /// <summary>
         /// Check responses; signal completion if we have received enough responses to determine outcome.
         /// </summary>
-        public void CheckIfDone()
+        private void CheckIfDone()
         {
             if (!Task.IsCompleted)
             {
-                if (responses.All(res => res != null && res.ResponseStatus == ActivationResponseStatus.Pass))
+                // store incomplete promises at this time (as they might be completed by the time the method finishes
+                var incompletePromises = responses.Where(t => !t.IsCompleted).ToArray();
+                if (incompletePromises.Length == 0 && responses.All(res => res.IsCompleted && res.Result.ResponseStatus == ActivationResponseStatus.Pass))
```

Comment:
I think checking the condition `res.IsCompleted` is redundant (because we already check `incompletePromises.Length ==0`)